### PR TITLE
E3SM upstream merge

### PR DESCRIPTION
**Pull request by tcclevenger**
Need the changes to COSP for mappy valgrind error.

### CODE REVIEW

**Review by jgfouca — DISMISSED** [review]
Any conflicts?

**Comment by brhillman** [comment]
Thanks for opening this. Have you tried running any SCREAM tests to see if anything breaks?

**Comment by tcclevenger** [comment]
Running the test suite now on my machine and all looks good so far.

**Comment by sarats** [comment]
Jim: Note that cnl.cmake is removed upstream, check if that impacts scream builds on Frontier.

**Comment by jgfouca** [comment]
@sarats , thanks for the heads-up, I forgot about that.

Looking at config_machines.xml, I do see one scream machine with OS=CNL:
```
  <machine MACH="frontier-scream-gpu">
    <DESC>Frontier. AMD EPYC 7A53 64C nodes, 128 hwthreads, 512GB DDR4, 4 MI250X GPUs.</DESC>
    <NODENAME_REGEX>.*frontier.*</NODENAME_REGEX>
    <OS>CNL</OS>
    ...
```

So this machine could be impacted by CNL removal.

Looking at the CNL.cmake settings:

```
if (NOT MPILIB STREQUAL impi)
  string(APPEND CMAKE_OPTS " -DCMAKE_SYSTEM_NAME=Catamount")
endif()
string(APPEND CPPDEFS " -DLINUX")
if (COMP_NAME STREQUAL gptl)
  string(APPEND CPPDEFS " -DHAVE_NANOTIME -DBIT64 -DHAVE_VPRINTF -DHAVE_BACKTRACE -DHAVE_SLASHPROC -DHAVE_COMM_F2C -DHAVE_TIMES -DHAVE_GETTIMEOFDAY")
endif()
set(MPICC "cc")
set(MPICXX "CC")
set(MPIFC "ftn")
set(MPI_LIB_NAME "mpich")
set(MPI_PATH "$ENV{MPICH_DIR}")
set(NETCDF_PATH "$ENV{NETCDF_DIR}")
set(PIO_FILESYSTEM_HINTS "lustre")
set(PNETCDF_PATH "$ENV{PARALLEL_NETCDF_DIR}")
set(SCC "cc")
set(SCXX "CC")
set(SFC "ftn")
```

and `crayclang-scream.cmake` and `crayclang-scream_frontier-scream-gpu.cmake`. We don't use mpi `impi`, so the first bit is not needed. 

We should add
`string(APPEND CPPDEFS " -DLINUX")`
to `crayclang-scream.cmake`.

We should add CPPDEFS for gptl as well to `crayclang-scream_frontier-scream-gpu.cmake`.

Setting mpi and serial compilers is not necessary (already being done in `crayclang-scream.cmake`).

Setting library paths is not necessary (already being done in `crayclang-scream_frontier-scream-gpu.cmake`).

The PIO_FILESYSTEM_HINTS needs to be added to `crayclang-scream_frontier-scream-gpu.cmake`.

@tcclevenger , I will add a commit to your PR and try it on frontier.

**Comment by tcclevenger** [comment]
Is there any reason to expect CIME diffs? Looks like all unit tests are passing.

**Comment by jgfouca** [comment]
@tcclevenger , can you look at the logs and see the details of the DIFFs?

**Comment by tcclevenger** [comment]
@jgfouca DIFFs in various SCREAM fields, ranging from around 1e-9 - 1e1 normalized difference.

**Comment by jgfouca** [comment]
@tcclevenger , that's not good... Can you jump on mappy and confirm it's this PR? Checkout the common ancestor between this branch and master, generate your own set of baselines, then checkout this branch and compare against those baselines.

**Comment by sarats** [comment]
@jgfouca Please take a look at the upstream PR to see if anything else needs to be updated. I think you have it mostly covered but take a closer look at https://github.com/E3SM-Project/E3SM/pull/5745/files

IIRC, I don't think 'PIO_FILESYSTEM_HINTS' is needed on Frontier. @jayeshkrishna?

**Comment by jgfouca** [comment]
@sarats , I'm not sure if it's needed, but CNL.cmake was adding it, so it should at least be consistent.

**Comment by jgfouca** [comment]
Neither master nor this branch builds on frontier with `crayclang-scream_frontier-scream-gpu`, so we aren't any worse off than before at least 🤷‍♂️

**Comment by sarats** [comment]
CNL.cmake was obsolete and so is that flag on machines where it doesn't have impact. I suggest leaving dead bits like that out for clarity. It's what I did on E3SM side.

**Comment by sarats** [comment]
You need to see impact after this PR is brought in: https://github.com/E3SM-Project/scream/pull/2503

**Comment by jgfouca** [comment]
@sarats , ok, I will merge your branch into this one and see what happens.

**Comment by jgfouca** [comment]
@sarats , still won't build. Cmake is unable to compile a simple program:

```
    Linking HIP executable cmTC_55a8a
    /autofs/nccs-svm1_sw/frontier/spack-envs/base/opt/linux-sles15-x86_64/gcc-7.5.0/cmake-3.21.3-rsmsjp6nq5d3ah55w4wrzjjcqgdzqmed/bin/cmake -E cmake_link_script CMakeFiles/cmTC_55a8a.dir/link.txt --verbose=1
    /opt/rocm-5.1.0/llvm/bin/clang++ --cuda-host-only  --offload-arch=gfx90a --hip-link CMakeFiles/cmTC_55a8a.dir/testHIPCompiler.hip.o -o cmTC_55a8a  /opt/rocm-5.1.0/lib/libamdhip64.so.5.1.50100 /opt/rocm-5.1.0/llvm/lib/clang/14.0.0/lib/linux/libclang_rt.builtins-x86_64.a 
    terminate called after throwing an instance of 'std::system_error'
      what():  Resource temporarily unavailable
```

Do I need to be a compute node to build?

**Comment by sarats** [comment]
No, you don't need a compute node.

Is this on Crusher? It's been throwing some weird errors.

**Comment by jgfouca** [comment]
@sarats , I'm on frontier:

```
% hostname -f
login10.frontier.olcf.ornl.gov
```

Crusher is not a `<OS>CNL</OS>` machine so cannot have been impacted by its removal.

**Comment by tcclevenger** [comment]
@jgfouca @brhillman So when I do a rebase now there are conflicts. I'm surprised they do not show up in the AT? Or is it necessarily that the have been introduced since I created the PR (which seems unlikely..)? What is the process of clearing them? I'm guessing it's not as simple as taking incoming changes.

**Comment by tcclevenger** [comment]
@jgfouca Ah good to know.. Ok, testing DIFFs on mappy. I'll report back.

**Comment by tcclevenger** [comment]
Looks like the DIFFs are real.

**Comment by jgfouca** [comment]
@tcclevenger , bummer. I wonder if CIME is responsible? Can you try setting the CIME submodule to what it was prior to the upstream merge, while leaving the rest of the upstream merge in tact, and see if the DIFF remains?

**Comment by tcclevenger** [comment]
@jgfouca Seems that updated CIME is necessary. Using older CIME I get
```
ERROR: Command: '/bin/xmllint --xinclude --noout --schema /home/tccleve/E3SM-Project/SCREAM/scream/cime/CIME/data/config/xml_schemas/config_archive.xsd /home/tccleve/E3SM-Project/SCREAM/scream/cime_config/config_archive.xml' failed with error '/home/tccleve/E3SM-Project/SCREAM/scream/cime_config/config_archive.xml:58: element comp_archive_spec: Schemas validity error : Element 'comp_archive_spec', attribute 'exclude_testing': The attribute 'exclude_testing' is not allowed.
        /home/tccleve/E3SM-Project/SCREAM/scream/cime_config/config_archive.xml:84: element comp_archive_spec: Schemas validity error : Element 'comp_archive_spec', attribute 'exclude_testing': The attribute 'exclude_testing' is not allowed.
        /home/tccleve/E3SM-Project/SCREAM/scream/cime_config/config_archive.xml:102: element comp_archive_spec: Schemas validity error : Element 'comp_archive_spec', attribute 'exclude_testing': The attribute 'exclude_testing' is not allowed.
        /home/tccleve/E3SM-Project/SCREAM/scream/cime_config/config_archive.xml:120: element comp_archive_spec: Schemas validity error : Element 'comp_archive_spec', attribute 'exclude_testing': The attribute 'exclude_testing' is not allowed.
        /home/tccleve/E3SM-Project/SCREAM/scream/cime_config/config_archive.xml fails to validate' from dir '/home/tccleve/E3SM-Project/SCREAM/scream'
```

I'll try and find where CIME was updated and test there.

**Comment by jgfouca** [comment]
@tcclevenger , try resetting the cime_config directory to pre-upstream.

**Comment by tcclevenger** [comment]
@jgfouca Looks like the diff came in before the CIME update. I'm bisecting now.

**Comment by tcclevenger** [comment]
Narrowed down to PR https://github.com/E3SM-Project/E3SM/pull/5781 as the culprit of the DIFFs. Description includes "Add remaining climate changing features for v3 atm", so this is probably expected?

**Comment by jgfouca** [comment]
@tcclevenger , nice work! That PR is giant, but I think you are right that this is expected.

**Comment by bartgol** [comment]
Wait, that PR changes stuff in EAM, and I don't see stuff that _we_ use. @mahf708 you were involved in that PR, can you comment on whether you would expect it to impact SCREAMv1 cases or not? The bisection suggests it does, but it's not clear to me why that's the case...

**Comment by tcclevenger** [comment]
@bartgol It is confusing, but I did test all the surrounding commits to be sure. It looks like `eam/src/physics/cam` is changed and `eam/src/chemistry`. I'm not familiar with chemistry part of the code at all, is there a chance some component touches that code?

**Comment by bartgol** [comment]
> @bartgol It is confusing, but I did test all the surrounding commits to be sure. It looks like `eam/src/physics/cam` is changed and `eam/src/chemistry`. I'm not familiar with chemistry part of the code at all, is there a chance some component touches that code?

I don't think we use eam's chemisty at all. As for `physics/cam`, I don't see any file change that makes me think it will affect EAMxx. Maybe @AaronDonahue or @brhillman have a better grasp.

**Comment by PeterCaldwell** [comment]
@crterai is probably the best person to comment on recent changes from the EAMf90 side which might affect EAMxx in this upstream merge. I doubt anything in chemistry or eam would impact us though.

**Comment by bartgol** [comment]
> @crterai is probably the best person to comment on recent changes from the EAMf90 side which might affect EAMxx in this upstream merge. I doubt anything in chemistry or eam would impact us though.

But we are seeing non-bfb change due to that atm-v3 PR. From what I can tell, it's a change in the default for CCSM_CO2_PPMV, which appears to be used in elm/mpas-ocean. I think it may be a benign diff, given what @mahf708 said, but since we're about to start a heavy prod campaign, we should be extra sure.

**Comment by bartgol** [comment]
> I would just reopen #2506 as-is and merge it in SCREAM

I would not reopen that PR, but rather do a new one which cherry picks the commits for COSP from the upstream. Cherry-picked commits will never generate conflicts, while doing "the same change" on both might (there's always a whitespace difference somewhere).

**Comment by PeterCaldwell** [comment]
> The EAMf90 merge above includes turning ON all v3 features. A lot of climate-changing stuff, so we should assume the default atmosphere coming from EAMf90 is distinctly different before/after this PR

But EAMxx autotester and nightlies don't run any of the v3 stuff, right? I'll be really bummed out if E3SM master screwed up our SCREAMv0 configuration.

**Comment by tcclevenger** [comment]
@bartgol I'll revert that parameter change and test.

**Comment by tcclevenger** [comment]
@bartgol Good catch, that parameter was the culprit. Commenting out that line returns us to BFB.

**Comment by jgfouca** [comment]
Just chiming in to say I don't think we should give up on this PR yet. I would be strongly in favor of seeing if we can adjust parameters to get back to BFB or at least understand the non-BFB. We need to go through the pain of this upstream merge at some point unless we want to permanently diverge from E3SM. @tcclevenger and I have already invested significant effort in getting this PR through.

**Comment by bartgol** [comment]
@PeterCaldwell Conrad confirmed that our v1 result change b/c the default for `CCSM_CO2_PPMV` for F2010 compsets has changed in the upstream. I don't want to stop the merge of this PR, but I think we _must_ confirm that the change in that parmaeter does not produce completely different climates. I think @mahf708 already said it's kind of a "fix" (the current value is more representative of a F2000 compset, IIUC), so I think this non-bfb change is ok. But you, @crterai , and any other eval folks should prob weigh in. Perhaps we should run a few days with this PR merged and check the climatology?

**Comment by tcclevenger** [comment]
@mahf708 It's all the tests in the `e3sm_scream_v1_lowres` group in `cime_config/tests.py`, and they all fail with DIFFs to our baseline output nc file. No test in EAMxx fail, because I believe nothing there uses this parameter, it's an issue in runs coupled with `elm/mpas-ocean`, which this parameter is passed (please correct if I'm wrong, @bartgol).

File: driver-mct/cime_config/config_component_e3sm.xml
Comment (jgfouca) [reply driver-mct/cime_config/config_component_e3sm.xml:731]: It will match any compset that starts with `2010`. F2010-SCREAMv1 is just an alias for a much longer full name that does start with 2010:

```
  <compset>
      <alias>F2010-SCREAMv1</alias> <!-- Coupled land-atm compset (using SCREAMv1), 2010 initial conditions -->
      <lname>2010_SCREAM_ELM%SPBC_CICE%PRES_DOCN%DOM_SROF_SGLC_SWAV_SIAC_SESP</lname>
      <support_level>Experimental, under development</support_level>
  </compset>
```

Comment (jgfouca) [reply driver-mct/cime_config/config_component_e3sm.xml:731]: Understandable. I think `<values match="last">` means go down the list of all these values and take the last one that matches. For `^2010`, this is regex syntax and `^` is a special character that refers to the start of the string.

Comment (bartgol) [reply driver-mct/cime_config/config_component_e3sm.xml:731]: Side topic: I was thinking about `^` and I don't understand how `F2010` is a match, given that `2010` is not at the start of the string...

Comment (bartgol) [reply driver-mct/cime_config/config_component_e3sm.xml:731]: Nvm, @mahf708 reminded me about longnames.

**Comment by tcclevenger** [comment]
@mahf708 This diff shows up also when comparing the `lnd_in` file:
```
BASE: co2_ppmv = 379.0
COMP: co2_ppmv = 388.717
```

**Comment by bartgol** [comment]
> > This diff shows up when comparing the lnd_in file:
> 
> > BASE: co2_ppmv = 379.0
> > COMP: co2_ppmv = 388.717
> 
> @bishtgautam (you're the one to come to mind): Do you know how the `co2_ppmv` parameter impacts lnd in an isolated fashion?
> 
> As Susannah noted ([E3SM-Project/E3SM#5799 (comment)](https://github.com/E3SM-Project/E3SM/pull/5799#issuecomment-1626317584)) this parameter does impact the HES (fka BGC) group, which was also my understanding that it would become an issue for things like deposition, etc., but only in an interactive fashion. It would be good to get to the bottom of what's causing the diff here so that we are better aware of it going forward. It may have gone under the cover of v3 (climate-changing darkness) in the integration...

If one really wants to figure out what changed, we could bisect: hack the buildnml code to force the old value of the param, once for elm only and a second time for mpas-ocean only. If they both diff, then the param impacts both lnd and ocn, otherwise only one of the component is impacted.

**Comment by PeterCaldwell** [comment]
Yeah, I'm fine with `CCSM_co2_ppmv` increasing from 379 to 388.7. It's a small change and probably makes the model more realistic. 

I am kind a weirded out that the land model would be getting a difference CO2 value than the atm - it seems like there should be a single CO2 parameter for the whole model. Do we know whether the atmospheric CO2 value is also 388.7? This isn't something that should stop our upstream merge, but seems like something the v3 team should be on top of. @golaz and @crterai - do you know if land and atm (and ocn!) are seeing the same CO2 values?

**Comment by crterai** [comment]
I recently ran an F2010 test case with the v3alpha02 version of the code and both  atm and lnd models are seeing `388.717` based on atm_in and lnd_in.

**Comment by AaronDonahue** [comment]
@tcclevenger , it's been two weeks since we last commented on this PR.  Is there anything more we need to do before we can merge this?

**Comment by tcclevenger** [comment]
@AaronDonahue Yeah, should be since we have the CESS branch. I should probably retest, and I can also look and see if any additional changes went into E3SM.

**Comment by AaronDonahue** [comment]
@tcclevenger , do we expect this to be non-BFB in the CIME based tests?  It seems to be DIFF'ing on mappy.

**Comment by tcclevenger** [comment]
@AaronDonahue Yes, 2 weeks ago we tested and confirmed it was only due to the `co2_ppmv` parameter change.

**Comment by AaronDonahue** [comment]
Is this ready to merge now?  It looks like the DIFFs we got are expected.

@tcclevenger , @PeterCaldwell @crterai , @bartgol , @jgfouca , @mahf708 , @bishtgautam ?

Pinging all of you because you commented on this PR in the past.

**Comment by bishtgautam** [comment]
The PR should be marked 'non-BFB'. It is good to go, as far as I'm concerned.

**Comment by tcclevenger** [comment]
@AaronDonahue Yes, good to go from my side. I'll let others comment if they have objections.

**Comment by jgfouca** [comment]
No objection here.

**Pull request merged**
